hazelcast/src/main/java/com/hazelcast/internal/nio/IOUtil.java

```diff
@@ -128,11 +128,26 @@ public static void writeObject(ObjectDataOutput out, Object object) throws IOExc
     public static <T> T readObject(ObjectDataInput in) throws IOException {
         boolean isBinary = in.readBoolean();
         if (isBinary) {
-            return (T) in.readData();
+            return (T) readData(in);
         }
         return in.readObject();
     }
 
+    public static void writeData(ObjectDataOutput out, Data data) throws IOException {
+        assert out instanceof DataWriter : "out must be an instance of DataWriter";
```

Comment:
can there be any cost of assertions even we disable it? Since before this PR, we have no assertions inside `writeData`.

Reply:
Did a benchmark performing `serializationService.toData()` on an `EntryEventData` instance with `Data` key & value (200 bytes long) that shows no slow down:

```
without assertions (commit )
CompletableFutureBenchmark.assertionCost  thrpt   50  2940496,668 ± 175423,573  ops/s

with assertions in writeData, assertions not enabled in JVM
CompletableFutureBenchmark.assertionCost  thrpt   50  3112858,606 ± 165626,511  ops/s
```